src/Middleware/RetryMiddleware.php

```diff
@@ -86,14 +94,17 @@ public function __invoke(Call $call, array $options)
         }
 
         return $nextHandler($call, $options)->then(null, function ($e) use ($call, $options) {
-            if (!$e instanceof ApiException) {
-                throw $e;
-            }
+            $retryFunction = $this->getRetryFunction();
 
-            if (!in_array($e->getStatus(), $this->retrySettings->getRetryableCodes())) {
+            // If the number of retries has surpassed the max allowed retries,
+            // or if the retry function returns false then throw the
+            // exception as we normally would.
+            if ($this->retryAttempts >= $this->retrySettings->getMaxRetries() ||
+                !$retryFunction($e, $options)) {
```

Comment:
when `maxRetries` is `0`, this represents infinite retries. So this logic needs to be updated for that case. 

Also, we can keep these as two separate cases, for readability
```suggestion
            if (0 !== $this->retrySettings->getMaxRetries()
                && $this->retryAttempts >= $this->retrySettings->getMaxRetries()
            ) {
                throw $e;
            }
            if (!$retryFunction($e, $options)) {
                throw $e;
            }
```